src/Functions/DateTimeTransforms.h

```diff
@@ -111,19 +111,30 @@ struct ToStartOfDayImpl
     //TODO: right now it is hardcoded to produce DateTime only, needs fixing later. See date_and_time_type_details::ResultDataTypeMap for deduction of result type example.
     static inline UInt32 execute(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
     {
-        return time_zone.toDate(static_cast<time_t>(t.whole));
+        if (t.whole < 0 || (t.whole >= 0 && t.fractional < 0))
+            return 0;
+
+        return time_zone.toDate(std::min(t.whole, time_t(0xffffffff)));
     }
     static inline UInt32 execute(UInt32 t, const DateLUTImpl & time_zone)
     {
         return time_zone.toDate(t);
     }
     static inline UInt32 execute(Int32 d, const DateLUTImpl & time_zone)
     {
-        return time_zone.toDate(ExtendedDayNum(d));
+        if (d < 0)
+            return 0;
+
+        Int64 date_time = time_zone.fromDayNum(ExtendedDayNum(d));
```

Comment:
`fromDayNum()` returns type `Time` ... we should use this typedef (or `auto`)

(same above)